### PR TITLE
Enables ob-bricks connector to read current IP(s) from the host

### PR DESCRIPTION
**Pull request by rahman-lb**
This code change enables ob-bricks connector  to read current IP(s) from the underlying host and pass this list to cinder driver.
 - Read available all available IP(s)
 - Discard local IP (127.0.0.1) from the list
 - Add this IPs list to connector properties along with hostNQN sent to cinder driver 
 -

### CODE REVIEW

File: os_brick/initiator/connectors/lightos.py [hunk 1]
```diff
@@ -62,6 +62,23 @@ def __init__(self,
         self.message_queue = message_queue
         self.DISCOVERY_DIR_PATH = '/etc/discovery-client/discovery.d/'
 
+    @staticmethod
+    def get_ip_addresses():
+        ip_addresses = []
+        result = subprocess.check_output(['ip', 'addr']).decode('utf-8')
```
Comment (muliby-lb) [inline os_brick/initiator/connectors/lightos.py:68]: I am not sure this will work, unfortunately. 
1. What guarantees do you have that `ip(1)` is installed? Does openstack require this package? If not, it needs to be added as a requirement, and the openstack maintainers may not agree to add it.
2. The openstack security model is annoying and we had to jump through a lot of hoops to do some things. Look at the `priv_lightos` stuff to see how we dealt with it. This may or may not be relevant here, but since the user can use `ip` to change the system IP address, I wouldn't be terribly surprised if openstack stops you from calling it without elevated privileges.
3. What is the format of `ip addr` output changes? Using binaries in scripts is dangerous when you want a long term solution. Better to use fixed system APIs, if possible.

File: os_brick/initiator/connectors/lightos.py [hunk 2]
```diff
@@ -62,6 +62,23 @@ def __init__(self,
...
+        lines = result.split('\n')
+        for line in lines:
+          if 'inet ' in line:
+              parts = line.split()
+              ip_parts = parts[1]
+              ip_parts = ip_parts.split('/')
+              host_ip=ip_parts[0]
+              if '127.0.0.1' in host_ip:
+                 continue
+              ip_addresses.append(host_ip)
```
Comment (muliby-lb) [inline os_brick/initiator/connectors/lightos.py:78]: I am not sure about the security implications here -- can you make a convincing argument that all IP addresses the host has (even if internal or tunnel or VPN or ...) should be added to the volume IPACL? I think if you look at the ticket you will see a list of addresses @sagi-lb recommended to filter out.

File: os_brick/initiator/connectors/lightos.py [hunk 3]
```diff
@@ -71,7 +88,9 @@ def get_connector_properties(root_helper, *args, **kwargs):
                                              execute=kwargs.get('execute'))
         hostnqn = utils.get_host_nqn()
         found_dsc = lightos_connector.find_dsc()
-
+        host_ips = lightos_connector.get_ip_addresses()
+        LOG.info('Host current IP(s) are ',host_ips)
```
Comment (muliby-lb) [inline os_brick/initiator/connectors/lightos.py:92]: Add which host, it will help debugging later.

**Review by muliby-lb — COMMENTED** [review]
On the plus side, this is a pretty simple implementation. On the negative side, I'm not sure it will work :-) Please see internal comments.

File: os_brick/initiator/connectors/lightos.py (hunk 2 above)
Comment (rahman-lb) [reply os_brick/initiator/connectors/lightos.py:78]: This returns all IPv4 addresses and writes to volume IP-ACL,  also return IPs when VPN is connected,  I tested on my local machine when i connect via VPN/tunnel to lab environment. I've not tested IPv6 and need to add support for  IPv6.

File: os_brick/initiator/connectors/lightos.py (hunk 1 above)
Comment (rahman-lb) [reply os_brick/initiator/connectors/lightos.py:68]: I think all modern Linux OS(s) are pre-packaged with command [ ip addr ] , but now i'm hesitant after you mentioned the security concerns.  Can you rephrase you question in point # 3     **What is the format of ip addr output changes?** 

I'm new to the system API.  Can you explain this     **Better to use fixed system APIs, if possible.**

@muliby-lb

Comment (muliby-lb) [reply os_brick/initiator/connectors/lightos.py:68]: Sorry, (3) should have been "what *if* the format changes". I am not sure it's a real concern but at the same time, getting the IP addresses from python, which calls the system APIs, seems pretty simple:

```
import socket

def get_system_ip_addresses():
    ip_addresses = []
    try:
        hostname = socket.gethostname()
        ip_addresses.append(socket.gethostbyname(hostname))
        ip_addresses.extend(socket.gethostbyname_ex(hostname)[-1])
    except Exception as e:
        print(f"An error occurred: {e}")
    return ip_addresses

if __name__ == "__main__":
    system_ip_addresses = get_system_ip_addresses()
    print("System IP addresses:", system_ip_addresses)
```

File: os_brick/initiator/connectors/lightos.py (hunk 2 above)
Comment (muliby-lb) [reply os_brick/initiator/connectors/lightos.py:78]: See @sagi-lb comment here https://lightbitslabs.atlassian.net/browse/LBM1-28409?focusedCommentId=113743 for which addresses to filter

File: os_brick/initiator/connectors/lightos.py (hunk 1 above)
Comment (rahman-lb) [reply os_brick/initiator/connectors/lightos.py:68]: I had tried this approach  but this was giving very limited IPs   e.g  these IPs on my machine 
['127.0.1.1', '127.0.1.1'] 

Then i switched to use [ ip add] which gives all available network ips 
['127.0.0.1', '10.0.0.115', '172.17.0.1', '172.18.0.1', '10.212.1.4']

IP-add  return both IPv4 and IPv6 address.

File: os_brick/initiator/connectors/lightos.py (hunk 2 above)
Comment (rahman-lb) [reply os_brick/initiator/connectors/lightos.py:78]: I'll apply filters

File: os_brick/initiator/connectors/lightos.py (hunk 1 above)
Comment (yuval-lb) [reply os_brick/initiator/connectors/lightos.py:68]: yea - this probably a problem, we somehow check the discovery client is running. I will check how we do that

Comment (muliby-lb) [reply os_brick/initiator/connectors/lightos.py:68]: @rahman-lb here's a better way to do what we need in pure python using the `netifaces` package:
https://stackoverflow.com/a/274644

File: os_brick/initiator/connectors/lightos.py (hunk 2 above)
Comment (muliby-lb) [reply os_brick/initiator/connectors/lightos.py:78]: I hope that using `netifaces` (see previous comment) it will be easier

File: os_brick/initiator/connectors/lightos.py (hunk 1 above)
Comment (yuval-lb) [reply os_brick/initiator/connectors/lightos.py:68]: I see that in connector.py there is a my_ip
wrote checking if that IP is good for us.

alternatives:
1. if we want to use ip addr its need to be in the privileged module 
2. see how the python way gives the correct addr
3. upload a patch to upstream and see of someone gives a better option

Comment (yuval-lb) [reply os_brick/initiator/connectors/lightos.py:68]: Also see this:
https://github.com/openstack/oslo.utils/blob/cecf061e6e4f666104170e3bba93fdae24255309/oslo_utils/netutils.py#L369

you can import it and use it like we do here:
https://github.com/LightBitsLabs/os-brick/blob/7dd8e0fdb828e5020f5c67a3e727e25990125a4b/os_brick/privileged/lightos.py#L18

File: os_brick/initiator/connectors/lightos.py (hunk 2 above)
Comment (rahman-lb) [reply os_brick/initiator/connectors/lightos.py:78]: netifaces works perfectly to get IPv4 and IPv6 address, trying to find if an address is physical or virtual for the filtering purposes.

Comment (yuval-lb) [reply os_brick/initiator/connectors/lightos.py:78]: I highly suggest using existing openstack tools like:
https://github.com/openstack/oslo.utils/blob/cecf061e6e4f666104170e3bba93fdae24255309/oslo_utils/netutils.py#L369

Comment (muliby-lb) [reply os_brick/initiator/connectors/lightos.py:78]: FWIW I 100% agree with @yuval-lb . If there is an openstack utility that does what we need, it's best to use it. It will make upstream submission much easier.

File: os_brick/initiator/connectors/lightos.py (hunk 3 above)
Comment (rahman-lb) [reply os_brick/initiator/connectors/lightos.py]: @muliby-lb @yuval-lb 
Interestingly netutil.py  is also using package netifaces, which means we should not have issues using netifaces. 
I'm using netutil.py for IPv4/6 validity checks. Current code will read all available IPs and pass-on to cinder. 
If IPv6 is not enabled on host then this function  will read only read IPv4 addresses.  
It will not be a bad idea to move forward with all IPs and then apply filtering on undesired IPs once everything works. 

@sagi-lb  Is there a way to determine if an IP belongs to a physical interface vs virtual.

File: os_brick/initiator/connectors/lightos.py (hunk 2 above)
Comment (rahman-lb) [reply os_brick/initiator/connectors/lightos.py]: @muliby-lb @yuval-lb
Interestingly netutil.py is also using package netifaces, which means we should not have issues using netifaces.
I'm using netutil.py for IPv4/6 validity checks. Current code will read all available IPs and pass-on to cinder.
If IPv6 is not enabled on host then this function will read only read IPv4 addresses.
It will not be a bad idea to move forward with all IPs and then apply filtering on undesired IPs once everything works.

@sagi-lb Is there a way to determine if an IP belongs to a physical interface vs virtual.

Comment (rahman-lb) [reply os_brick/initiator/connectors/lightos.py]: oslo_utils/netutils.py has function that returns only one IP,  but not all IPs. In a multi IPs environment this will become a problem.

File: os_brick/initiator/connectors/lightos.py [hunk 4]
```diff
@@ -64,20 +66,39 @@ def __init__(self,
...
+        loop_back_ips=['127.0.0.1','0:0:0:0:0:0:0:1', '::1']
+        ips = []
+        is_ipv6_enabled = netutils.is_ipv6_enabled()
+                          
+        ifaces = interfaces()
+        for iface in ifaces:
+            addresses = ifaddresses(iface)
+            
+            for address_family in (AF_INET, AF_INET6):
+                family_addresses = addresses.get(address_family)
```
Comment (muliby-lb) [inline os_brick/initiator/connectors/lightos.py:81]: `addresses.get(address_family, None)` unless you are absolutely sure that `addresses` always has the `address_family`.

File: os_brick/initiator/connectors/lightos.py [hunk 5]
```diff
@@ -64,20 +66,39 @@ def __init__(self,
...
+                                        
+                    if is_ipv6_enabled and netutils.is_valid_ipv6(ip):
+                       parts = ip.split("%")
+                       ip = parts[0]
+                       ips.append(ip)
+
+                    elif netutils.is_valid_ipv4(ip):
+                         ips.append(ip)
+                    else:
+                        LOG.info('Not a usable ip address  %s', ip)
```
Comment (muliby-lb) [inline os_brick/initiator/connectors/lightos.py:99]: This will spam the log like crazy. Should not go into production.

**Comment by ronen-lb** [comment]
Closing as superseded: this change shipped via #8 (`yoga/os-brick-ip-acl`, merged 2023-11-28) and landed upstream as opendev commit `837bd71` ("add a new funtion to return IP address(es) of a host machine"), first released in os-brick 6.9.0 / OpenStack 2024.2 Dalmatian.

Context: while investigating LBM1-47891 (new cinder `lightos.py` requires the `host_ips` connector property, which pre-Dalmatian os-brick — e.g. RHOSO 18 / Antelope — does not provide, breaking attachment), we traced the provenance of the `host_ips` change and found this draft PR still open.

cc @yuval-lb @rahman-lb
